File: tests/fixtures/go/store/store.go
// Package store is a struct with methods, so that `parent` has something to
// disagree with the file about.
//
// Every method here is declared at file scope. Nothing lexically encloses
// them; only the receiver says who owns them. That is the case docs/plan.md
// M5 names by hand.
package store

import (
	"fmt"
	str "strings"
)

// Entry is one entry in the store.
type Entry struct {
	Value string
}

// Store is a tiny key-value store.
type Store struct {
	entries map[string]Entry
}

// Get reads one entry.
//
// Two lines of documentation, to prove newlines survive.
func (s *Store) Get(key string) (Entry, bool) {
	found, ok := s.entries[key]
	return found, ok
}

// Put writes one entry. A value receiver where Get took a pointer one, so the
// owner is found through both spellings.
func (s Store) Put(key string, value string) {
	s.entries[key] = Entry{Value: value}
}

func (s *Store) evict() {
	clear(s.entries)
}

// Describe renders an entry. A second type with methods in the same file, so
// that resolving a receiver to the wrong one would show up.
func (e Entry) Describe() string {
	return fmt.Sprintf("%s", str.ToUpper(e.Value))
}

// Warm is a free function that calls a method, so `calls` has an edge to find.
func Warm(s *Store) (Entry, bool) {
	return s.Get("warm")
}

File: tests/fixtures/rust/src/store.rs
//! The store: a struct with an inherent impl, so that `parent` has something
//! to disagree with the file about.

use std::collections::HashMap;
use std::fmt::Debug as Show;

/// An entry in the store.
#[derive(Debug, Clone)]
pub struct Entry {
    pub value: String,
}

/// A tiny key-value store.
pub struct Store {
    entries: HashMap<String, Entry>,
}

impl Store {
    /// Read one entry.
    ///
    /// Two lines of documentation, to prove newlines survive.
    pub fn get(&self, key: &str) -> Option<Entry> {
        let found = self.entries.get(key);
        found.cloned()
    }

    /// Write one entry.
    #[inline]
    pub fn put(&mut self, key: String, value: String) {
        self.entries.insert(key, Entry { value });
    }

    fn evict(&mut self) {
        self.entries.clear();
    }
}

/// A free function that calls a method, so `calls` has an edge to find.
pub fn warm(store: &Store) -> Option<Entry> {
    store.get("warm")
}

mod detail {
    /// Nested in a module, so the descriptor chain is more than one hop.
    pub fn helper() -> u32 {
        7
    }
}

fn show<T: Show>(_value: T) {}
